e3db/tests/test_tsv1_auth.py

```diff
@@ -0,0 +1,95 @@
+
+def test_private_signing_key_decode_and_encode():
+    '''
+    Asserts that decoding and re-encoding does not alter the private
+    signing key. 
+    '''
+    private_b64_decoded = BaseCrypto.base64decode(CREDS["private_signing_key"])
+    encoded = BaseCrypto.base64encode(private_b64_decoded).decode("utf-8")
+    assert(encoded == CREDS["private_signing_key"])
+
+def test_public_signing_key_decode_and_encode():
+    '''
+    Asserts that decoding and re-encoding does not alter the public
+    signing key. 
+    '''
+    public_b64_decoded = BaseCrypto.base64decode(CREDS["public_signing_key"])
+    encoded = BaseCrypto.base64encode(public_b64_decoded).decode("utf-8")
+    assert(encoded == CREDS["public_signing_key"])
+
+def test_public_signing_key_derivation():
+    '''
+    Asserts that deriving the public signing key from the PyNacl SigningKey 
+    is equivalent to the known public signing key. 
+    '''
+    private_b64_decoded = BaseCrypto.base64decode(CREDS["private_signing_key"])
+    signing_key = SodiumCrypto.generate_signing_key(private_b64_decoded[:SECRET_KEY_BYTES])
+    derived_public_signing_key = signing_key.verify_key
+    public_signing_key_encoded = BaseCrypto.base64encode(derived_public_signing_key).decode("utf-8")
+    assert(public_signing_key_encoded == CREDS["public_signing_key"])
+
+def test_private_signing_key_derivation():
+    '''
+    Asserts that deriving the private signing key from the PyNacl SigningKey
+    is equivalent to the known private signing key. 
+    '''
+    private_b64_decoded = BaseCrypto.base64decode(CREDS["private_signing_key"])
+    signing_key = SodiumCrypto.generate_signing_key(private_b64_decoded[:SECRET_KEY_BYTES])
+    derived_private_signing_key = signing_key._signing_key
+    private_signing_key_encoded = BaseCrypto.base64encode(derived_private_signing_key).decode("utf-8")
+    assert(private_signing_key_encoded == CREDS["private_signing_key"])
+
+def test_blake2b_hash():
+    '''
+    Asserts that the Blake2B hash of the string to sign is equivalent to the
+    known hash value of the string to sign. 
+    '''
+    string_to_hash =  f"{CALL_PATH}; {QUERY_STRING}; {CALL_METHOD}; {HEADER_STRING}"
+    hash_to_sign = BaseCrypto.hashString(string_to_hash).hex()
+    print(f"HASH TO SIGN: {hash_to_sign}")
+    assert(hash_to_sign == KNOWN_HASH)
+
+def test_signature():
+    '''
+    Asserts that the generated Ed25519 signature is equivalent to the known signature value. 
+    '''
+    string_to_hash =  f"{CALL_PATH}; {QUERY_STRING}; {CALL_METHOD}; {HEADER_STRING}"
+    string_to_sign = BaseCrypto.hashString(string_to_hash)
+    private_b64_decoded = BaseCrypto.base64decode(CREDS["private_signing_key"])[:SECRET_KEY_BYTES]
+    full_signature = SodiumCrypto.sign_string(string_to_sign, private_b64_decoded)
+    signature_64 = BaseCrypto.base64encode(full_signature.signature).decode("utf-8")
+    assert(signature_64 == KNOWN_SIGNATURE)
```

Comment:
I think you would be well served with a test that actually passes in a `request` object and verifies that the header is correct. Given that time and nonce are needed refactoring to a testable method that takes those as parameters might be a good idea. These tests make a number of assumptions about the shape of fields within the `request` object.

Reply:
Is there another way to get a `request` object besides manually with `requests.Request()`? Looking at its definition it looks like its shape will be unchanged from the parameters I give, so another option I see is to take in the request and make it a `PreparedRequest` with `r.prepare()` for the test case? That object has a method `prepare_url()` that should check the params, I believe.

Reply:
You can manually create a request using its init method. `requests.Request(method=None, url=None, headers=None,files=None, data=None,params=None, auth=None, cookies=None, hooks=None, json=None)`
The open questions that need to be answered from my perspective are

- is r.url the same as the call path as defined in the other sdks (I don't believe it is, I am pretty sure it is a full URL. TSV1 only uses the path after the domain
- is r.params when stringified always going to be sorted and in the format we expect. https://github.com/psf/requests/blob/f6d43b03fbb9a1e75ed63a9aa15738a8fce99b50/requests/models.py#L211 the library seems to be able to stringify it before making the request but it also seems to support maps and lists of tuples.
- is the timestamp that is generated from datetime.datetime.now() down the the second or millisecond? Is that different once stringified?. Does that match the scheme that TSV1 has?

There might be more but that is the starting point.